src/fiber_util/fiber_util.ml

```diff
@@ -102,3 +102,37 @@ module Cancellation = struct
             Cancelled x
           | Not_cancelled -> Fiber.return Not_cancelled)
 end
+
+module Svar = struct
+  type 'a t =
+    { mutable current : 'a
+    ; mutable waiters : ('a Fiber.Ivar.t * ('a -> bool)) list
+    }
+
+  let read t = t.current
+
+  let wait t ~until =
```

Comment:
Shouldn't this function return immediately if the `until` predicate is satisfied? The current behaviour seems a bit surprising.

Reply:
Indeed. I fixed it. I had it this way, because I thought that the polling loop would be written by doing `read` and then looping with `until`. But we can just use always use `until` in this way.

IMO we should just formalize the pattern and leave users less room for error. Something like this perhaps:

```ocaml
type 'a loop
val await : 'a t -> eq:('a -> 'a -> bool) -> 'a loop
val next : 'a loop -> 'a Fiber.t
```

Reply:
The current `Svar` API is simple and expressive -- it lets you model classic state machines with accepting states (expressed with the `until` predicate). You can build all sorts of things on top of it, including these `await`/`next` primitives, with an appropriate choice of `'a`, e.g.:

```ocaml
type 'a with_previous = { previous : 'a option; current : 'a }
```

What kind of errors do you think users are going to make with the current API?

Reply:
We don't need "all sorts of things". In practice we only have one pattern:

```ocaml
let rec loop last =
  Poller.on_request poller >>= function
  | None -> Fiber.return ()
  | Some () ->
    let* v = Svar.wait svar ~until:(fun a -> Some a = last) in
    let* () = Poller.respond poller v in
    loop (Some v)
in loop None
```

Or do you foresee this module being useful elsewhere?

Reply:
This seems like a fine use-case for the current API, although I guess you meant `Some a <> last` in the predicate?

Given a choice between something very specific (like the version with `await` and `next`, tailored for this use-case) and something simple and generic (like `Svar`), I tend to lean towards the latter.

It seems much easier to explain the semantics of `wait ~until` compared to `await` which immediately introduces complications like the distinction between the previous and the current states, what the "previous state" means in the initial state, what the `'a loop` type means, and so on.

Reply:
I guess I don't consider the Svar construction useful enough to warrant generalization beyond this single use case. By the way, `'a loop` can be replaced with some fold as well. In any case, I already wrote tests for this implementation let's move on to other things.